src/Npgsql/TaskTimeoutAndCancellation.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.Util;
+
+namespace Npgsql;
+
+static class TaskTimeoutAndCancellation
+{
+    /// <summary>
+    /// Utility that executes a non-cancellable task while allowing to timeout and/or cancel awaiting for it.
+    /// If the given task does not complete within <paramref name="timeout"/>, a <see cref="TimeoutException"/> is thrown.
+    /// </summary>
+    /// <param name="getTaskFunc">Gets a <see cref="Task{TResult}"/>.</param>
+    /// <param name="timeout">The timeout after which the <see cref="Task"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+    /// <param name="useLegacyImplementation">An optional <see cref="Boolean"/> parameter for testing only. If set to true the method uses pre-dotnet 6.0 implementation for cancellation and timeout enforcement.</param>
+    /// <typeparam name="TResult">The result <see cref="Type"/>.</typeparam>
+    /// <returns>The <see cref="Task{TResult}"/> representing the asynchronous wait.</returns>
+    internal static async Task<TResult> WaitAsync<TResult>(Func<CancellationToken, Task<TResult>> getTaskFunc, NpgsqlTimeout timeout, CancellationToken cancellationToken, bool useLegacyImplementation = false)
```

Comment:
Maube be a silly question: does the repo runs the test pipelines for all supported .NET platforms (.e. 3.1 ... 6.0)?

Reply:
Removed `useLegacyImplementation`

Reply:
@baal2000 it doesn't currently, as we haven't had any issues with older TFMs - but that would be quite easy to do.

Reply:
I was afraid that is the case and added `useLegacyImplementation` flag because of that. Otherwise I would not be comfortable commiting new code to the repository knowing the legacy TFMs are not regression-tested.

> that would be quite easy to do

Could you make a plan for that?

Reply:
@baal2000 I fully understand, have opened #4258 to track this - I'll take a look at it.